File: data-structures/Algorithms/quickSort.py
def partition(nums, low, high):
    pivot = nums[low]

    i, j = low, high
    while i < j:
        while nums[i] <= pivot and i<=high-1:
            i += 1
        while nums[j] > pivot and j>= low+1:
            j -= 1
        if i < j:
            nums[i], nums[j] = nums[j], nums[i]
    nums[low], nums[j] = nums[j], nums[low]
    return j

def quickSort(nums, low, high):
    if low >= high:
        return
    # find partition index
    partitionIndex = partition(nums, low, high)
    # call recursively on two subarrays
    quickSort(nums, low, partitionIndex-1)
    quickSort(nums, partitionIndex+1, high)

nums = [5,2,3,3,5,4,1]
quickSort(nums, 0 ,len(nums)-1)
print(nums)

nums2 = [324,1342,354,34,34,3243,545,0]
quickSort(nums2, 0, len(nums2)-1)
print(nums2)

File: algorithms-and-sums/Sorting/quickSort.cpp
#include<bits/stdc++.h>

using namespace std;

int partition(vector<int> &arr, int low, int high){
	int left = low;
	int right = high;
	int pivot = arr[low];
	
	while(left < right){
		while(arr[left] <= pivot && left<=high-1){
			left++;
		}
		while(arr[right] > pivot && right>=low+1){
			right--;
		}
		if(left < right) {
			int temp = arr[left];
			arr[left] = arr[right];
			arr[right] = temp;
		}
		int temp = arr[right];
		arr[right] = arr[low];
		arr[low] = temp;
	}
	return right;
}

void quickSort(vector<int> &arr, int left, int right){
	if(left > right){
		return;
	}
	int partitionIndex = partition(arr, left, right);
	quickSort(arr, left, partitionIndex-1);
	quickSort(arr, partitionIndex+1, right);
}

template <typename T> void print(vector<T> arr){
    for(auto num: arr){
        cout << num << " ";
    }
    cout << endl;
}

int main(){
    vector<int> arr = {1,0,5,2,4,5,26,2,63,66,3,0,45,2,3};
    quickSort(arr, 0, arr.size()-1);
    print(arr);
}